XBOXONE.cpp

```diff
@@ -0,0 +1,359 @@
+                }
+#endif
+        return rcode;
+}
+
+void XBOXONE::readReport() {
+        if(readBuf == NULL)
+                return;
+        if(readBuf[0] == 0x07) {
+                // The XBOX button has a seperate message
+                if (readBuf[4] == 1) {
+                        ButtonState |= XBOX_BUTTONS[XBOX];
+                }
+                else {
+                        ButtonState &= XBOX_BUTTONS[XBOX];
+                }
+                if(ButtonState != OldButtonState) {
+                        ButtonClickState = ButtonState & ~OldButtonState; // Update click state variable, but don't include the two trigger buttons L2 and R2
+                        OldButtonState = ButtonState;
+                }
+        }
+        if(readBuf[0] != 0x20) { // Check if it's the correct report, otherwise return - the controller also sends different status reports
+#ifdef EXTRADEBUG
+                Notify(PSTR("\r\nXbox Poll: "), 0x80);
+                D_PrintHex<uint8_t > (readBuf[0], 0x80); // 0x03 is a heart beat report!
+#endif
+                return;
+        }
+
+        uint16_t xbox = ButtonState & XBOX_BUTTONS[XBOX]; // since the XBOX button is seperate, save it and add it back in
+        // xbox button from before, dpad, abxy, start/back, sync, stick click, shoulder buttons
+        ButtonState = xbox | (((uint16_t)readBuf[5] & 0xF) << 8) | (readBuf[4] & 0xF0)  | (((uint16_t)readBuf[4] & 0x0C) << 10) | ((readBuf[4] & 0x01) << 3) | (((uint16_t)readBuf[5] & 0xC0) << 8) | ((readBuf[5] & 0x30) >> 4);
+
+        triggerValue[LeftTrigger] = (uint16_t)(((uint16_t)readBuf[7] << 8) | readBuf[6]);
+        triggerValue[RightTrigger] = (uint16_t)(((uint16_t)readBuf[9] << 8) | readBuf[8]);
+
+        hatValue[LeftHatX] = (int16_t)(((uint16_t)readBuf[11] << 8) | readBuf[10]);
+        hatValue[LeftHatY] = (int16_t)(((uint16_t)readBuf[13] << 8) | readBuf[12]);
+        hatValue[RightHatX] = (int16_t)(((uint16_t)readBuf[15] << 8) | readBuf[14]);
+        hatValue[RightHatY] = (int16_t)(((uint16_t)readBuf[17] << 8) | readBuf[16]);
+
+        //Notify(PSTR("\r\nButtonState"), 0x80);
+        //PrintHex<uint16_t>(ButtonState, 0x80);
+
+        if(ButtonState != OldButtonState) {
+                ButtonClickState = ButtonState & ~OldButtonState; // Update click state variable
+                OldButtonState = ButtonState;
+        }
+}
+
+void XBOXONE::printReport() { //Uncomment "#define PRINTREPORT" to print the report send by the Xbox ONE Controller
+#ifdef PRINTREPORT
+        if(readBuf == NULL)
+                return;
+        for(uint8_t i = 0; i < XBOX_REPORT_BUFFER_SIZE; i++) {
+                D_PrintHex<uint8_t > (readBuf[i], 0x80);
+                Notify(PSTR(" "), 0x80);
+        }
+        Notify(PSTR("\r\n"), 0x80);
+#endif
+}
+
+uint8_t XBOXONE::getButtonPress(ButtonEnum b) {
+        return (bool)(ButtonState & ((uint16_t)pgm_read_word(&XBOX_BUTTONS[(uint8_t)b])));
+}
+
+bool XBOXONE::getButtonClick(ButtonEnum b) {
```

Comment:
Could you check for L2/R2 here, as it is done in the XBOXRECV code: https://github.com/felis/USB_Host_Shield_2.0/blob/master/XBOXRECV.cpp#L418-L430. I know it looks ugly, but it should be okay for now. I am thinking of rewriting a lot of the code at some point anyway, then I will fix stuff like this as well.